File: packages/runtime/src/durable-protection.ts
/**
 * The guarded `events/v1alpha2` write and read path for the durable scheduler.
 *
 * spec/redaction-semantics.md Section 7 requires that every default runtime sink
 * receive data only after the shared deterministic guard has completed, and that
 * "the default runtime dependency graph exposes only sinks whose public write
 * method accepts a `PreparedSinkWrite`". This module is the durable scheduler's
 * single door to the journal: it turns a scheduler event draft into one guarded
 * record whose authoritative application values exist on the wire only as
 * validated `ProtectedValueRef`s, and it resolves those references back into
 * detached values on recovery.
 *
 * There is no fallback. If a compatible `ProtectedPayloadStore` and
 * `KeyProvider` are not configured, `assertPayloadProtection` fails with
 * `PAYLOAD_PROTECTION_REQUIRED` before the first event, checkpoint, log, error
 * payload, temporary file, or executor invocation (Section 4.2).
 *
 * Honest claim boundary: a recovered value is decoded into process memory so the
 * scheduler can bind it as node input. Section 13 already states that this
 * contract does not hide values "from an authorized executor after decryption or
 * from the runtime process memory needed to execute". What it does guarantee is
 * that no byte of that value reaches a sink outside the protected blob.
 */

import { canonicalHash } from "@graph-engineering/core";
import {
  DEFAULT_CAPTURE_POLICY,
  PROTECTED_STORE_CONTRACT,
  ProtectedEventReader,
  SinkGuard,
  canonicalTagged,
  computeValueMac,
  hmacSha256Hex,
  keyRefHash,
  prepareProtectedEvent,
  type AuthorityScope,
  type CapturePolicy,
  type CaptureSinkClass,
  type CaptureSourceClass,
  type GraphEventType,
  type GraphEventV1Alpha2,
  type GraphEventV1Alpha2Type,
  type GuardFailure,
  type KeyProvider,
  type PreparedSinkWrite,
  type ProtectedEventSpec,
  type ProtectedPayloadStore,
  type SemanticContext,
  type SinkScannerOptions,
} from "@graph-engineering/persistence";

import { decodeDurableJson } from "./durable-json.js";
import { DurableRunError, type DurableRunErrorCode } from "./durable-types.js";
import type { JsonValue } from "./types.js";

export const DURABLE_CONTRACT_VERSION = "scheduler-recovery/v1alpha2" as const;
export const DURABLE_GRAPH_REVISION = 1;

/**
 * The guarded journal surface. `ProtectedJsonlEventStore` and
 * `MemoryProtectedEventStore` both satisfy it structurally. The only way to put
 * a byte in one is a `PreparedSinkWrite` minted by the guard for that exact
 * instance, so an adapter cannot opt out by calling a lower-level raw writer.
 */
export interface GuardedDurableJournal {
  readonly sink: CaptureSinkClass;
  readonly binding: object;
  append(
    runId: string,
    expectedVersion: number,
    writes: readonly PreparedSinkWrite[],
  ): Promise<number>;
  read(runId: string, fromSequence?: number): AsyncIterable<GraphEventV1Alpha2>;
}

/** Operator-owned payload protection for one durable run (Sections 4.1 and 5.3). */
export interface DurablePayloadProtection {
  readonly journal: GuardedDurableJournal;
  readonly payloadStore: ProtectedPayloadStore;
  readonly keys: KeyProvider;
  readonly scope: AuthorityScope;
  /** Defaults to the Section 4.1 stable profile. */
  readonly policy?: CapturePolicy;
  readonly scanner?: SinkScannerOptions;
}

export interface DurablePayloadDraft {
  /** The `data` member that will hold the reference. */
  readonly field: "inputRef" | "outputRef" | "resultRef" | "evidenceRef";
  readonly semanticContext: SemanticContext;
  /** The raw application value. It never enters the candidate record. */
  readonly value: unknown;
}

export interface DurableEventDraft {
  readonly type: GraphEventV1Alpha2Type;
  /** Closed metadata members only. No application value belongs here. */
  readonly data: Readonly<Record<string, unknown>>;
  readonly payloads?: readonly DurablePayloadDraft[];
  readonly nodeId?: string;
  readonly edgeId?: string;
  readonly attempt?: number;
}

/**
 * One persisted event with its protected references resolved back into detached
 * values. This is a recovery projection held in process memory; it is never
 * serialized to any sink.
 */
export interface RecoveredEvent {
  readonly eventId: string;
  readonly type: GraphEventType;
  readonly timestamp: string;
  readonly runId: string;
  readonly graphRevision: number;
  readonly sequence: number;
  readonly nodeId?: string;
  readonly edgeId?: string;
  readonly attempt?: number;
  readonly payloadHash: string;
  readonly data: Readonly<Record<string, unknown>>;
}

/**
 * Section 1.2 source classification for each scheduler event type. A name that
 * sounds like metadata does not make its contents safe, so every event carrying
 * an application payload is classified by what that payload actually is.
 */
const EVENT_SOURCE_CLASS: Readonly<Record<GraphEventV1Alpha2Type, CaptureSourceClass>> =
  Object.freeze({
    RunCreated: "graph-input",
    RunStarted: "runtime-generated-identifier",
    RunResumed: "runtime-generated-identifier",
    NodeScheduled: "bound-node-input",
    NodeStarted: "runtime-generated-identifier",
    NodeAttemptFailed: "exception-message",
    NodeSettledWithoutAttempt: "node-result",
    NodeRetried: "runtime-generated-identifier",
    NodeSucceeded: "node-output",
    EdgeEmitted: "runtime-generated-identifier",
    RunCancelled: "run-result",
    RunFailed: "run-result",
    RunSucceeded: "run-result",
  });

const MAC_FIELD: Readonly<Record<DurablePayloadDraft["field"], string>> = Object.freeze({
  inputRef: "inputMac",
  outputRef: "outputMac",
  resultRef: "resultMac",
  evidenceRef: "evidenceMac",
});

/** Section 10 codes are portable; they surface as durable run error codes. */
function protectionCode(failure: GuardFailure): DurableRunErrorCode {
  switch (failure.code) {
    case "PAYLOAD_PROTECTION_REQUIRED":
    case "PAYLOAD_PROTECTION_FAILED":
    case "CAPTURE_POLICY_MISMATCH":
    case "INLINE_CAPTURE_NOT_AUTHORIZED":
    case "PROTECTED_PAYLOAD_NOT_FOUND":
    case "PROTECTED_PAYLOAD_UNAUTHORIZED":
    case "PROTECTED_PAYLOAD_CORRUPT":
    case "SECRET_CANARY_DETECTED":
      return failure.code;
    default:
      return "PAYLOAD_PROTECTION_FAILED";
  }
}

/**
 * Section 10: the failure is structured and does not echo the offending value.
 * Only the stable code, phase, safe reason token, and run/node identifiers are
 * carried out.
 */
export function protectionError(runId: string, failure: GuardFailure): DurableRunError {
  return new DurableRunError(
    protectionCode(failure),
    runId,
    "durable payload protection refused the operation",
    {
      code: failure.code,
      phase: failure.phase,
      reason: failure.reason,
      ...(failure.sink === undefined ? {} : { sink: failure.sink }),
      ...(failure.sourceClass === undefined ? {} : { sourceClass: failure.sourceClass }),
    },
  );
}

/**
 * Section 4.2: a durable start, resume, or checkpoint save that would persist an
 * authoritative application value without a compatible `ProtectedPayloadStore`
 * plus `KeyProvider` fails with `PAYLOAD_PROTECTION_REQUIRED` before the first
 * event, checkpoint, log, error payload, temporary plaintext file, or executor
 * invocation. It never generates a key beside the ciphertext and never falls
 * back to inline capture.
 */
export function assertPayloadProtection(
  runId: string,
  protection: DurablePayloadProtection | undefined,
): asserts protection is DurablePayloadProtection {
  const missing: string[] = [];
  const has = (owner: unknown, ...methods: readonly string[]): boolean =>
    typeof owner === "object" &&
    owner !== null &&
    methods.every((name) => typeof (owner as Record<string, unknown>)[name] === "function");

  if (typeof protection !== "object" || protection === null) {
    missing.push("protection");
  } else {
    if (!has(protection.journal, "append", "read")) missing.push("journal");
    if (!has(protection.payloadStore, "put", "get")) missing.push("payloadStore");
    if (
      !has(protection.keys, "protectionKey", "runIdentityKey", "nonce") ||
      typeof protection.keys.keyRef !== "string" ||
      protection.keys.keyRef.length === 0
    ) {
      missing.push("keys");
    }
    const scope = protection.scope as unknown as Record<string, unknown> | undefined;
    if (
      typeof scope !== "object" ||
      scope === null ||
      typeof scope["tenantScopeId"] !== "string" ||
      typeof scope["authorityProviderId"] !== "string" ||
      typeof scope["authoritySubjectId"] !== "string"
    ) {
      missing.push("scope");
    }
  }
  if (missing.length === 0) return;
  throw new DurableRunError(
    "PAYLOAD_PROTECTION_REQUIRED",
    runId,
    "durable persistence requires a protected payload store and key provider",
    { missing, contract: PROTECTED_STORE_CONTRACT },
  );
}

/** Section 8.2 activity/idempotency key. Attempt is deliberately excluded. */
export function durableActivityKey(
  identityKey: Uint8Array,
  fields: { runId: string; graphRevision: number; nodeId: string; inputMac: string },
): string {
  return hmacSha256Hex(
    identityKey,
    canonicalTagged([
      "activity/v1alpha2",
      fields.runId,
      fields.graphRevision,
      fields.nodeId,
      fields.inputMac,
    ]),
  );
}

interface ProtectedFieldSpec {
  readonly field: DurablePayloadDraft["field"];
  readonly semanticContext: SemanticContext;
  /** The hydrated `data` member that receives the decoded value. */
  readonly target: string;
  /** True when the hydrated member holds the tagged encoding, not the value. */
  readonly tagged: boolean;
}

/**
 * One protected durable run: the guard, the guarded journal it is bound to, and
 * the authorized reader for its references.
 */
export class ProtectedDurableRun {
  readonly runId: string;
  readonly #protection: DurablePayloadProtection;
  readonly #guard: SinkGuard;
  readonly #reader: ProtectedEventReader;
  readonly #identityKey: Uint8Array;

  constructor(protection: DurablePayloadProtection, runId: string) {
    this.runId = runId;
    this.#protection = protection;
    this.#guard = new SinkGuard({
      policy: protection.policy ?? DEFAULT_CAPTURE_POLICY,
      keys: protection.keys,
      store: protection.payloadStore,
      scope: protection.scope,
      ...(protection.scanner === undefined ? {} : { scanner: protection.scanner }),
    });
    this.#reader = new ProtectedEventReader({
      keys: protection.keys,
      store: protection.payloadStore,
      scope: protection.scope,
      capturePolicyHash: this.#guard.capturePolicyHash,
    });
    this.#identityKey = protection.keys.runIdentityKey(runId);
  }

  get capturePolicyHash(): string {
    return this.#guard.capturePolicyHash;
  }

  get keyRefHash(): string {
    return keyRefHash(this.#protection.keys.keyRef);
  }

  /** Section 4.4: a resume whose policy hash differs is `CAPTURE_POLICY_MISMATCH`. */
  assertPolicyHash(expected: unknown): void {
    if (typeof expected !== "string") {
      throw new DurableRunError(
        "CAPTURE_POLICY_MISMATCH",
        this.runId,
        "durable run does not record a capture policy hash",
      );
    }
    const failure = this.#guard.assertPolicyHash(expected);
    if (failure !== undefined) throw protectionError(this.runId, failure);
  }

  /** Section 5.4 semantic value MAC, the logical identity used in recovery. */
  valueMac(semanticContext: SemanticContext, value: unknown): string {
    return computeValueMac(this.#identityKey, semanticContext, value);
  }

  activityKey(nodeId: string, inputMac: string): string {
    return durableActivityKey(this.#identityKey, {
      runId: this.runId,
      graphRevision: DURABLE_GRAPH_REVISION,
      nodeId,
      inputMac,
    });
  }

  /**
   * Guard one batch and commit it as a unit. Section 7: a multi-event commit is
   * rejected as a unit before a dependent is released, so every prepared write
   * is minted before the journal is touched.
   */
  async append(
    expectedVersion: number,
    drafts: readonly DurableEventDraft[],
    eventIdFor: (sequence: number, type: GraphEventV1Alpha2Type) => string,
    timestamp: () => string,
  ): Promise<number> {
    const prepared: PreparedSinkWrite[] = [];
    for (const [index, draft] of drafts.entries()) {
      const sequence = expectedVersion + 1 + index;
      const eventId = eventIdFor(sequence, draft.type);
      const spec: ProtectedEventSpec = {
        decisionId: `${this.runId}.${sequence}.${draft.type}`,
        eventId,
        type: draft.type,
        timestamp: timestamp(),
        runId: this.runId,
        graphRevision: DURABLE_GRAPH_REVISION,
        sequence,
        sourceClass: EVENT_SOURCE_CLASS[draft.type],
        data: draft.data,
        ...(draft.nodeId === undefined ? {} : { nodeId: draft.nodeId }),
        ...(draft.edgeId === undefined ? {} : { edgeId: draft.edgeId }),
        ...(draft.attempt === undefined ? {} : { attempt: draft.attempt }),
        ...(draft.payloads === undefined || draft.payloads.length === 0
          ? {}
          : {
              payloads: draft.payloads.map((payload) => ({
                field: payload.field,
                macField: MAC_FIELD[payload.field],
                semanticContext: payload.semanticContext,
                value: payload.value,
              })),
            }),
      };
      const result = await prepareProtectedEvent(this.#guard, this.#protection.journal, spec);
      if (result.kind === "failed") throw protectionError(this.runId, result.failure);
      if (result.kind === "suppressed") {
        throw protectionError(this.runId, {
          code: "PAYLOAD_PROTECTION_FAILED",
          phase: "policy",
          reason: "capture-policy-suppressed-an-authoritative-durable-event",
        });
      }
      prepared.push(result.prepared);
    }
    return this.#protection.journal.append(this.runId, expectedVersion, prepared);
  }

  /** Read the persisted stream with every protected reference resolved. */
  async read(): Promise<readonly RecoveredEvent[]> {
    const persisted: GraphEventV1Alpha2[] = [];
    for await (const event of this.#protection.journal.read(this.runId)) persisted.push(event);
    const recovered: RecoveredEvent[] = [];
    for (const event of persisted) recovered.push(await this.#recover(event));
    return recovered;
  }

  #fields(event: GraphEventV1Alpha2): readonly ProtectedFieldSpec[] {
    const runId = event.runId;
    const graphRevision = event.graphRevision;
    const nodeId = event.nodeId as string;
    switch (event.type) {
      case "RunCreated":
        return [{
          field: "inputRef",
          semanticContext: { kind: "graph-input", runId, graphRevision },
          target: "input",
          tagged: true,
        }];
      case "NodeScheduled":
        return [{
          field: "inputRef",
          semanticContext: { kind: "node-input", runId, graphRevision, nodeId },
          target: "input",
          tagged: true,
        }];
      case "NodeSucceeded":
        return [{
          field: "outputRef",
          semanticContext: { kind: "node-output", runId, graphRevision, nodeId },
          target: "output",
          tagged: true,
        }];
      case "NodeSettledWithoutAttempt":
        return [{
          field: "resultRef",
          semanticContext: { kind: "node-result", runId, graphRevision, nodeId },
          target: "result",
          tagged: true,
        }];
      case "RunSucceeded":
      case "RunFailed":
      case "RunCancelled":
        return [{
          field: "resultRef",
          semanticContext: { kind: "run-result", runId, graphRevision },
          target: "result",
          tagged: true,
        }];
      case "NodeAttemptFailed": {
        if (event.data["evidenceRef"] === undefined) return [];
        const inline = event.data["failure"];
        const code =
          typeof inline === "object" && inline !== null
            ? String((inline as Record<string, unknown>)["code"])
            : "";
        return [{
          field: "evidenceRef",
          semanticContext: {
            kind: "diagnostic-evidence",
            runId,
            graphRevision,
            nodeId,
            attempt: event.attempt as number,
            code,
          },
          target: "failure",
          tagged: false,
        }];
      }
      default:
        return [];
    }
  }

  async #recover(event: GraphEventV1Alpha2): Promise<RecoveredEvent> {
    // The recovery projection keeps every closed metadata member and every
    // adjacent `*Mac` identity, drops the opaque references it just resolved,
    // and drops `evidenceMac` because the resolved evidence replaces the closed
    // failure projection wholesale.
    const data: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(event.data)) {
      if (key.endsWith("Ref") || key === "evidenceMac") continue;
      data[key] = value;
    }

    for (const spec of this.#fields(event)) {
      const resolved = await this.#reader.resolve(event, spec.field, spec.semanticContext);
      if (!resolved.ok) throw protectionError(this.runId, resolved.failure);
      let tagged: unknown;
      try {
        tagged = JSON.parse(resolved.taggedJson);
      } catch {
        throw protectionError(this.runId, {
          code: "PROTECTED_PAYLOAD_CORRUPT",
          phase: "encode",
          reason: "protected-plaintext-is-not-tagged-json",
        });
      }
      if (spec.tagged) {
        data[spec.target] = tagged;
        continue;
      }
      let decoded: JsonValue;
      try {
        decoded = decodeDurableJson(tagged);
      } catch {
        throw protectionError(this.runId, {
          code: "PROTECTED_PAYLOAD_CORRUPT",
          phase: "encode",
          reason: "protected-plaintext-is-not-canonical-durable-json",
        });
      }
      // Section 6.1 keeps the closed inline projection authoritative for the
      // fields it carries; the protected evidence must agree with it exactly.
      const inline = data[spec.target];
      if (typeof inline === "object" && inline !== null && typeof decoded === "object" &&
          decoded !== null && !Array.isArray(decoded)) {
        for (const [key, value] of Object.entries(inline as Record<string, unknown>)) {
          if (JSON.stringify((decoded as Record<string, unknown>)[key]) !== JSON.stringify(value)) {
            throw protectionError(this.runId, {
              code: "PROTECTED_PAYLOAD_CORRUPT",
              phase: "classification",
              reason: "protected-evidence-contradicts-its-closed-metadata-projection",
            });
          }
        }
      }
      data[spec.target] = decoded;
    }

    return {
      eventId: event.eventId,
      type: event.type,
      timestamp: event.timestamp,
      runId: event.runId,
      graphRevision: event.graphRevision,
      sequence: event.sequence,
      ...(event.nodeId === undefined ? {} : { nodeId: event.nodeId }),
      ...(event.edgeId === undefined ? {} : { edgeId: event.edgeId }),
      ...(event.attempt === undefined ? {} : { attempt: event.attempt }),
      payloadHash: canonicalHash(data),
      data,
    };
  }
}

File: python/src/graph_engineering/durable_protection.py
"""Operator-owned payload protection for a durable run.

``redaction-semantics.md`` Section 4.2 is the whole point of this module: if a
durable start or checkpoint would persist an authoritative application value and
a compatible ``ProtectedPayloadStore`` plus ``KeyProvider`` are not configured,
the operation fails with ``PAYLOAD_PROTECTION_REQUIRED`` before the first event,
checkpoint, log, error payload, temporary plaintext file, or executor
invocation.  There is deliberately no no-op key provider, no in-process default
store, and no fallback to the legacy inline writer, because each of those would
turn a refusal into silent plaintext on disk.

A :class:`PayloadProtection` bundles the three things the durable runtime needs
and cannot invent for itself: the key authority, the protected blob store, and
the immutable effective capture policy.  From them it builds the one shared
:class:`~graph_engineering.redaction.guard.SinkGuard` every durable sink write
passes through, and it provides the inverse operation the recovery fold needs to
turn a committed protected reference back into an application value.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Final

from .canonical import canonical_sha256
from .models import JsonValue
from .redaction.errors import RedactionFailure, failure
from .redaction.guard import SinkGuard
from .redaction.keys import KeyProvider, Protector
from .redaction.limits import DEFAULT_LIMITS, PortableLimits
from .redaction.policy import CapturePolicy, capture_policy_hash, default_stable_profile
from .redaction.protect import (
    ProtectedAad,
    ProtectedPayloadStore,
    ProtectedValueRef,
    authority_binding_hash,
    key_ref_hash,
    tenant_scope_hash,
    unprotect_value,
    validate_reference_document,
)
from .redaction.scan import CanaryRegistry

#: The transform implementation this package ships. Section 4.1 requires the
#: effective policy to name one; it identifies the pointer-transform code, and it
#: is never derived from source or result bytes.
TRANSFORM_IMPLEMENTATION_MANIFEST: Final = (
    "graph-engineering/python/redaction/json-pointer-rules/v1alpha2"
)
TRANSFORM_IMPLEMENTATION_HASH: Final = canonical_sha256(
    ["transform-implementation/v1alpha2", TRANSFORM_IMPLEMENTATION_MANIFEST]
)

#: The default profile enables no redaction rule at all, so the registry
#: snapshot this runtime resolves is the empty ordered registry at version 1.
RULE_REGISTRY_VERSION: Final = 1
RULE_REGISTRY_HASH: Final = canonical_sha256(["rule-registry/v1alpha2", RULE_REGISTRY_VERSION, []])


def default_durable_capture_policy(key_ref: str) -> CapturePolicy:
    """The Section 4.1 default stable profile pinned to one key reference."""

    return default_stable_profile(
        transform_implementation_hash=TRANSFORM_IMPLEMENTATION_HASH,
        rule_registry_version=RULE_REGISTRY_VERSION,
        rule_registry_hash=RULE_REGISTRY_HASH,
        key_ref=key_ref,
    )


class PayloadProtection:
    """The configured protection authority for one durable runtime.

    Construction requires a real key provider and a real protected payload
    store.  Neither has a default, so "unconfigured" is representable only as
    the absence of this object, which the durable entry points reject.
    """

    __slots__ = (
        "_authority_provider_id",
        "_authority_subject_id",
        "_guard",
        "_key_provider",
        "_limits",
        "_payload_store",
        "_policy",
        "_policy_hash",
        "_protector",
        "_tenant_scope_id",
    )

    def __init__(
        self,
        *,
        key_provider: KeyProvider,
        payload_store: ProtectedPayloadStore,
        policy: CapturePolicy | None = None,
        protector: Protector | None = None,
        tenant_scope_id: str = "tenant-local",
        authority_provider_id: str = "authority-local",
        authority_subject_id: str = "subject-local",
        canaries: CanaryRegistry | None = None,
        limits: PortableLimits = DEFAULT_LIMITS,
    ) -> None:
        if not isinstance(key_provider, KeyProvider):
            raise TypeError("payload protection requires a KeyProvider")
        if not isinstance(payload_store, ProtectedPayloadStore):
            raise TypeError("payload protection requires a ProtectedPayloadStore")
        resolved = policy if policy is not None else default_durable_capture_policy(
            key_provider.key_ref
        )
        if resolved.key_ref != key_provider.key_ref:
            raise ValueError("capture policy keyRef does not match the configured key provider")
        self._key_provider = key_provider
        self._payload_store = payload_store
        self._policy = resolved
        self._policy_hash = capture_policy_hash(resolved)
        self._protector = protector
        self._tenant_scope_id = tenant_scope_id
        self._authority_provider_id = authority_provider_id
        self._authority_subject_id = authority_subject_id
        self._limits = limits
        self._guard = SinkGuard(
            policy=resolved,
            key_provider=key_provider,
            store=payload_store,
            tenant_scope_id=tenant_scope_id,
            authority_provider_id=authority_provider_id,
            authority_subject_id=authority_subject_id,
            protector=protector,
            canaries=canaries,
            limits=limits,
        )

    @property
    def guard(self) -> SinkGuard:
        return self._guard

    @property
    def key_provider(self) -> KeyProvider:
        return self._key_provider

    @property
    def payload_store(self) -> ProtectedPayloadStore:
        return self._payload_store

    @property
    def policy(self) -> CapturePolicy:
        return self._policy

    @property
    def policy_hash(self) -> str:
        return self._policy_hash

    @property
    def key_ref_digest(self) -> str:
        return key_ref_hash(self._key_provider.key_ref)

    def identity_key(self, run_id: str) -> bytes:
        """The Section 5.3 run identity key, stable across resume."""

        return self._key_provider.run_identity_key(run_id)

    def reference_from_document(
        self, document: object
    ) -> ProtectedValueRef | RedactionFailure:
        """Validate one committed closed reference before it can affect state."""

        invalid = validate_reference_document(document)
        if invalid is not None:
            return invalid
        assert isinstance(document, Mapping)
        return ProtectedValueRef(
            ref=str(document["ref"]),
            ciphertext_hash=str(document["ciphertextHash"]),
            value_mac=str(document["valueMac"]),
            key_ref_hash=str(document["keyRefHash"]),
            aad_hash=str(document["aadHash"]),
        )

    def recover(
        self,
        reference: ProtectedValueRef,
        *,
        run_id: str,
        graph_revision: int,
        record_type: str,
        event_id: str,
        sequence: int,
        field_path: str,
        capture_policy_hash_value: str,
        semantic_context: Mapping[str, JsonValue],
        node_id: str | None = None,
        edge_id: str | None = None,
        attempt: int | None = None,
    ) -> JsonValue | RedactionFailure:
        """Authenticate one committed reference under its exact event AAD.

        The AAD is rebuilt from the persisted occurrence rather than stored
        beside the reference, so a reference copied from another event, run,
        field path, or capture policy fails authentication before the fold can
        observe any value.
        """

        if reference.key_ref_hash != self.key_ref_digest:
            return failure("PROTECTED_PAYLOAD_UNAUTHORIZED", "protect", runId=run_id)
        identity_key = self.identity_key(run_id)
        tenant = tenant_scope_hash(identity_key, self._tenant_scope_id)
        aad = ProtectedAad(
            run_id=run_id,
            graph_revision=graph_revision,
            record_kind="event",
            record_type=record_type,
            sequence=sequence,
            field_path=field_path,
            capture_policy_hash=capture_policy_hash_value,
            key_ref_hash=self.key_ref_digest,
            authority_binding_hash=authority_binding_hash(
                identity_key,
                authority_provider_id=self._authority_provider_id,
                authority_subject_id=self._authority_subject_id,
                tenant_scope=tenant,
                run_id=run_id,
                capture_policy_hash=capture_policy_hash_value,
                key_reference_hash=self.key_ref_digest,
            ),
            tenant_scope_hash=tenant,
            value_mac=reference.value_mac,
            event_id=event_id,
            node_id=node_id,
            edge_id=edge_id,
            attempt=attempt,
        )
        return unprotect_value(
            reference,
            aad=aad,
            semantic_context=semantic_context,
            store=self._payload_store,
            key_provider=self._key_provider,
            protector=self._protector,
        )
